src/providers/MeetingProvider/MeetingManager.ts

```diff
@@ -110,6 +112,8 @@ export class MeetingManager implements AudioVideoObserver {
 
   simulcastEnabled: boolean = false;
 
+  videoDownlinkBandwidthPolicyWrapper: ((logger: Logger) => VideoDownlinkBandwidthPolicy) | null = null;
```

Comment:
@devalevenkatesh @simmkyu Could you please give me some feedback on this design?

To initialize a `VideoDowninkBandwidthPolicy`, we must pass in a `logger`.
```js
new VideoPriorityBasedPolicy(logger);
```
The React SDK is different from JS SDK. In React SDK, we create `logger` in the `initializeMeetingSession` in [`MeetingManager` here](https://github.com/aws/amazon-chime-sdk-component-library-react/blob/ae5874591d4895a182a826384a84bfe9ec378074/src/providers/MeetingProvider/MeetingManager.ts#L201).

```js
async initializeMeetingSession(): Promise<any> {
  const logger = this.createLogger(configuration);
  ...
```

There are two solutions to allow user pass a customized downlink policy.

### 1. Wrap the downlink policy as a callback, and trigger the callback after initializing the logger

Wrap the `VideoDowninkBandwidthPolicy` as a callback. Pass the downlink policy wrapper via `MeetingConfig`.

```js
// In meetingConfig.ts

...

const videoDownlinkBandwidthPolicyWrapper = (logger: Logger): VideoPriorityBasedPolicy => {
  return new VideoPriorityBasedPolicy(logger);
};

const config = {
  logLevel,
  postLogConfig,
  videoDownlinkBandwidthPolicyWrapper,
};
```
Trigger the Wrapper to initialize the downlink policy.

```js
async initializeMeetingSession(
  configuration: MeetingSessionConfiguration,
  deviceLabels: DeviceLabels | DeviceLabelTrigger = DeviceLabels.AudioAndVideo,
): Promise<any> {
  const logger = this.createLogger(configuration);

  if (this.videoDownlinkBandwidthPolicyWrapper) {
    configuration.videoDownlinkBandwidthPolicy = this.videoDownlinkBandwidthPolicyWrapper(logger);
  } 

  const deviceController = new DefaultDeviceController(logger);
  this.meetingSession = new DefaultMeetingSession(
    configuration,
    logger,
    deviceController,
    this.eventReporter
  );
```

#### pros
* No need to change the design of logger
* Only add a new property to `MeetingConfig`

#### cons
* The logic is not very straightforward, we need to wrap the downlink policy as a callback

### 2. Initialize the logger and downlink policy, pass them via MeetingConfig

```js
// In meetingConfig.ts

...

const logger = new ConsoleLogger('SDK', logLevel);
const videoDownlinkBandwidthPolicy = new VideoPriorityBasedPolicy(logger);

const config = {
  logLevel,
  postLogConfig,
  logger,
  videoDownlinkBandwidthPolicy,
};
```

```
// In the MeetingManager.ts

...

  constructor(config: ManagerConfig) {
    this.logLevel = config.logLevel;

    if (config.simulcastEnabled) {
      this.simulcastEnabled = config.simulcastEnabled;
    }

    if (config.postLogConfig) {
      this.postLoggerConfig = config.postLogConfig;
    }

    if (config.logger) {
      this.logger = config.logger;
    }
    
    if (config.videoDownlinkBandwidthPolicy) {
      this.videoDownlinkBandwidthPolicy = config.videoDownlinkBandwidthPolicy;
    }
  }

...

  async initializeMeetingSession(
    configuration: MeetingSessionConfiguration,
    deviceLabels: DeviceLabels | DeviceLabelTrigger = DeviceLabels.AudioAndVideo,
  ): Promise<any> {
    const logger = this.createLogger(configuration);

    if (this.videoDownlinkBandwidthPolicy) {
      configuration.videoDownlinkBandwidthPolicy = this.videoDownlinkBandwidthPolicy;
    }

    const deviceController = new DefaultDeviceController(logger);
    this.meetingSession = new DefaultMeetingSession(
      configuration,
      logger,
      deviceController,
      this.eventReporter
    );
```

#### pros
* Things are pretty straightforward, user can initiazlie the logger and downlink policy themselves and pass them via `MeetingConfig`

#### cons
* The newly added `logger` attribute conflicts with the existing (`logLevel` and `postLogConfig`), user should use either former or latter
* Need to add the logic to deal with this conflict

Which do you think is a better idea for this feature?

Reply:
Yes, I checked your posted solution and looks like first way is the one we can go with as of now. Getting the logger out would be good but confusing with the existing method as you pointed out and overall my opinion is we may anyhow require a refactor on how the meeting configuration is done currently to make it more flexible and easily extensible and if there are better patterns/ways to do this which may result in a v3 😅 .

Reply:
Based on the discussion, I will use the second solution.